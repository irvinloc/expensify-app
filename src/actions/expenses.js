import { v4 as uuidv4 } from 'uuid';
import moment from 'moment';
import database from '../firebase/firebase';


export const addExpense = (expense)=>({
    type: "ADD_EXPENSE",
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description='', 
            note='',
            amount=0,
            createdAt=moment().valueOf()
        } = expenseData;
        const expense = { description, note, amount, createdAt};
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id:ref.key, 
                ...expense
            }));
        });
    };
};

export const editExpense = (id, updates )=>({
        type: "EDIT_EXPENSE",
        expense: {
            id,
            updates
        }
    });
export const removeExpense = ( {id}={}) =>({
        type: "REMOVE_EXPENSE",
        id
    });

