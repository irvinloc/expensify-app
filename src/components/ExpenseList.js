import React from 'react';
import { connect } from 'react-redux';
import getVisibleExpenses from '../selectors/expenses';
import ExpenseListItem from './ExpenseListItem';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

 
export const ExpenseList = (props) => (
    
    <div>
        
        {props.expenses.length===0 ? (<p>Please, register some expenses</p>):
        (props.expenses.map(x=> <ExpenseListItem  key={x.id} {...x} />))}
    </div>
);

const mapStateToProps=(state)=>{
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters),
        expenses_total: selectExpensesTotal(getVisibleExpenses(state.expenses, state.filters))
    };
};
export default connect(mapStateToProps)(ExpenseList);
