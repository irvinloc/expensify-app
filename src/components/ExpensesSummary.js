import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

 
export const ExpensesSummary = ({expenseCount, expensesTotal}) => {
    const expenseWord = expenseCount ===1 ? 'expense': 'expenses';
    return (
    <div>
        <h2>Viewing {expenseCount} {expenseWord} 
        totalling { numeral(expensesTotal/100).format('$0,0.00').replace('$', 'â‚¬') }</h2>
    </div>

    );
}
const mapStateToprops = (state) =>
{
    const visibleExpenses=selectExpenses(state.expenses, state.filters);
    return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
    }
}

export default connect(mapStateToprops)(ExpensesSummary);