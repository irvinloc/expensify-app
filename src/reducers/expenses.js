const defaultExpensesReducerState=[];
export default (state=defaultExpensesReducerState, action) => {
    switch(action.type) {                
        case "ADD_EXPENSE" :
            return  [...state, action.expense]  
        case "EDIT_EXPENSE" :
            console.log('EDIT_EXPENSE');
            console.log('x.id',action.expense.id);
            console.log(action.expense);
            return state.map((x)=>x.id===action.expense.id? {...x, ...action.expense.updates}: x)
        case "REMOVE_EXPENSE" :
            return  state.filter((x)=>x.id!==action.id)           
        default: 
            return state
    }
};