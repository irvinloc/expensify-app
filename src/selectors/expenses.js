import moment from 'moment'
export default (expenses,{text, sortBy, startDate, endDate}) => {
    
    return expenses.filter((exp)=> {
        const createdAtMoment =  moment(exp.createdAt);
        
        const startDateMatch = startDate? startDate.isSameOrBefore(createdAtMoment, 'day'): true;
        const endDateMatch = endDate? endDate.isSameOrAfter(createdAtMoment, 'day'): true;
        
        const textmatch = !!!text || (exp.description.toLowerCase().includes(text.toLowerCase())) 
                          || (exp.note.toLowerCase().includes(text.toLowerCase())) 
        return startDateMatch && endDateMatch && textmatch
    }).sort((a,b)=> {
      if (!!!sortBy)  {
        return 0;
      } 
      else if (sortBy==='date')  { 
        return a.createdAt<b.createdAt?1:-1;
      } else if (sortBy==='amount') {
        return a.amount<b.amount?1:-1;
      }
    })
};