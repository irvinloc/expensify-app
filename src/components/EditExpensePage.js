import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {
    onSubmit=(expense) => {
        // props.dispatch(addExpense(expense));
        this.props.startEditExpense(this.props.expense.id, expense);
        
        this.props.history.push('/');
    };
    onRemoveExpense=() => {
        this.props.startRemoveExpense({id:this.props.expense.id});
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <h1>Edit Expense </h1>
                
                <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit}>
                </ExpenseForm>
                <button  onClick={this.onRemoveExpense} > DELETE </button>
                    
            </div>
        );
     }
}

const mapStoreToProps = (state, props) =>{
    return {
        expense: state.expenses.find(x=>x.id===props.match.params.id)
    }
}
const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id,expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id)=> dispatch(startRemoveExpense(id))
});


export default connect(mapStoreToProps, mapDispatchToProps)(EditExpensePage);