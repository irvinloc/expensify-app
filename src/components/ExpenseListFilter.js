import React from 'react';
import moment from 'moment';
import { connect } from 'react-redux';
import { DateRangePicker} from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate, resetFilters } from '../actions/filters';

// import 'react-dates/lib/css/_datepicker.css';

export class ExpenseListFilter extends React.Component {
    state = {
        calendarFocused: null
    };
    onDatesChange=({startDate, endDate})=> {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange= (calendarFocused) => { this.setState(()=>({ calendarFocused }));}
        
 
    onResetDates = ()=>{
        this.props.setStartDate(moment().add(-50, 'years'));
        this.props.setEndDate(moment().add(50, 'years'));
    }
    onResetAllFilters= ()=> {
        this.props.resetFilters();
    }
    onTextChange= (e) => this.props.setTextFilter(e.target.value);
    onSelectChange= (e)=> e.target.value==="date"?this.props.sortByDate():this.props.sortByAmount();
    
    onTextChanged=(e)=>{this.props.dispatch(setTextFilter(e.target.value));}
    render()  {
        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextChange}
                />
                <select 
                    onChange={this.onSelectChange} 
                    value={this.props.filters.sortBy}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                startDate={this.props.filters.starDate} 
                endDate={this.props.filters.endDate} 
                startDateId="your_unique_start_date_id"
                endDateId="your_unique_end_date_id"
                onDatesChange={this.onDatesChange} 
                focusedInput={this.state.calendarFocused} 
                onFocusChange={this.onFocusChange} 
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={()=>false}
                />
                <button onClick={this.onResetDates}>Reset Dates</button>
                <button onClick={this.onResetAllFilters}>Reset All Filters</button>
            </div>
        );
    }
};


const mapStateToProps=(state)=>({
    filters: state.filters
});
const mapDispatchToProps = (dispatch) => ({
    // DatesChange: ({startDate, endDate}) => {
    //     if (startDate) {dispatch(setStartDate(startDate));}
    //     if (endDate) {dispatch(setEndDate(endDate));}
    // },
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    resetFilters: ()=> dispatch(resetFilters())
    
});


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);
