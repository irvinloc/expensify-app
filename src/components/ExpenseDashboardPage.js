import React from 'react';
import ExpenseList from './ExpenseList'
import { XYPlot, XAxis, YAxis, VerticalGridLines, HorizontalGridLines, LineSeries } from 'react-vis';
import { connect } from 'react-redux';
// import { NavLink}  from 'react-router-dom';
import  getVisibleExpenses  from '../selectors/expenses'
import  getTotalExpenses  from '../selectors/expenses-total'
// import { setGenerating, resetGenerating }  from '../actions/filters'
import ExpenseListFilter from './ExpenseListFilter';
import ExpensesSummary from './ExpensesSummary';
import makeBins from '../utils/binsForHistogram';


export const ExpenseDashboardPage =(props)=> (
    <div>
  
    {/* <button onClick={(e)=>{ props.dispatch(setGenerating()) } }>GENERATE</button>
    <button onClick={(e)=>{ props.dispatch(resetGenerating()) }}>STOP GENERATE</button> */}
        <div>
            <ExpensesSummary/>
            
            <ExpenseListFilter  expenseCount={getTotalExpenses(getVisibleExpenses(props.expenses, props.filters))} expensesTotal= {props.expenses.length} />
        </div>
        <ExpenseList/>
        <div className="content-container">
        <XYPlot
            width={600}
            height={300}
            yDomain={[0,100]}
            animate= {true}
            >

            <HorizontalGridLines   />
            <LineSeries 
                data={makeBins(getVisibleExpenses(props.expenses, props.filters))}/>
            <XAxis />
            <YAxis />
        </XYPlot>
        </div>
        
    </div> 
);


const mapStateToProps=(state) =>{
    return {
        expenses:state.expenses,
        filters:state.filters
    }
}
export default connect(mapStateToProps)(ExpenseDashboardPage);

//export default ExpenseDashboardPage;