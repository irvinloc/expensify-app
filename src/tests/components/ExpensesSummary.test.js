import React from 'react';

import { shallow } from 'enzyme';
// import toJSON from 'enzyme-to-json';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses'; 


test('should render component correctly without expenses', () => {
    
    const wrapper = shallow(<ExpensesSummary expensesCount={0} expensesTotal={0}/>);
    expect(wrapper).toMatchSnapshot(); 
});


test('should render component correctly with one expense', () => {
    const expensesTotal=[expenses[1]].map(x=>x.amount).reduce((a,b)=>a+b);
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={expensesTotal}/>);
    expect(wrapper).toMatchSnapshot(); 
});


test('should render component correctly', () => {
    const expensesTotal=expenses.map(x=>x.amount).reduce((a,b)=>a+b);
    const wrapper = shallow(<ExpensesSummary expenseCount={expenses.length} expensesTotal={expensesTotal}/>);
    expect(wrapper).toMatchSnapshot(); 
});

