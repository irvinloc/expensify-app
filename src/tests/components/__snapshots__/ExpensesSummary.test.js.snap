// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component correctly 1`] = `
<div>
  <h2>
    Viewing 
    3
     
    expenses 
    totalling 
    €1,141.95
  </h2>
</div>
`;

exports[`should render component correctly with one expense 1`] = `
<div>
  <h2>
    Viewing 
    1
     
    expense 
    totalling 
    €1,095.00
  </h2>
</div>
`;

exports[`should render component correctly without expenses 1`] = `
<div>
  <h2>
    Viewing 
     
    expenses 
    totalling 
    €0.00
  </h2>
</div>
`;
